pkg/synchromanager/clustersynchro/apiservice_controller.go

```diff
@@ -0,0 +1,150 @@
+
+	reconciled     bool
+	reconcileCh    chan struct{}
+	discoveryCache discovery.GroupVersionCache
+
+	informer cache.SharedIndexInformer
+	lister   apiregistrationlisters.APIServiceLister
+}
+
+func NewAPIServiceController(cluster string, config *rest.Config, discoveryCache discovery.GroupVersionCache) (*APIServiceController, error) {
+	client, err := aggregatorclientset.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	informer := apiregistrationinformers.NewAPIServiceInformer(client, 0, nil)
+	controller := &APIServiceController{
+		cluster:        cluster,
+		discoveryCache: discoveryCache,
+		reconcileCh:    make(chan struct{}, 1),
+		informer:       informer,
+		lister:         apiregistrationlisters.NewAPIServiceLister(informer.GetIndexer()),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(_ interface{}) { controller.triggerReconcile() },
+		UpdateFunc: func(old, new interface{}) {
+			oldObj, oldOk := old.(*apiregistrationv1api.APIService)
+			newObj, newOk := old.(*apiregistrationv1api.APIService)
+			if !oldOk || !newOk {
+				return
+			}
+			if equality.Semantic.DeepEqual(oldObj.Spec, newObj.Spec) {
+				if oldObj.Spec.Service == nil || equality.Semantic.DeepEqual(oldObj.Status, newObj.Status) {
+					return
+				}
+			}
+			controller.triggerReconcile()
+		},
+		DeleteFunc: func(_ interface{}) { controller.triggerReconcile() },
+	})
+	return controller, nil
+}
+
+func (c *APIServiceController) triggerReconcile() {
+	select {
+	case c.reconcileCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *APIServiceController) Run(stopCh <-chan struct{}) {
+	go c.informer.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, c.informer.HasSynced)
+
+	select {
+	case <-c.reconcileCh:
+	default:
+	}
+	c.reconcile()
+	c.reconciled = true
+
+	for {
+		select {
+		case <-c.reconcileCh:
+		case <-stopCh:
+			return
+		}
+
+		select {
+		case <-stopCh:
+			return
+		default:
+		}
+		c.reconcile()
+	}
+}
+
+func (c *APIServiceController) HasSynced() bool {
+	return c.informer.HasSynced() && c.reconciled
+}
+
+func (c *APIServiceController) reconcile() {
+	apiservices, err := c.lister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "reconcile apiservices failed", "cluster", c.cluster)
+		return
+	}
+
+	groupVersions := make(map[string][]string)
+	aggregatorGroups := sets.NewString()
+
+	apiServicesByGroup := apiregistrationv1helper.SortedByGroupAndVersion(apiservices)
+	for _, groupServices := range apiServicesByGroup {
+		apiServicesByGroup := apiregistrationv1helper.SortedByGroupAndVersion(groupServices)[0]
+		group := apiServicesByGroup[0].Spec.Group
+
+		var versions []string
+		var isLocal, isAggregator bool
+		for _, apiService := range apiServicesByGroup {
+			if apiService.Spec.Service != nil {
+				isAggregator = true
+				if !apiregistrationv1helper.IsAPIServiceConditionTrue(apiService, apiregistrationv1api.Available) {
+					// ignore unavailable aggregator version
+					continue
+				}
```

Comment:
The condition state has not yet been updated when a new aggregation API is created, and this check ignores the new APIService.

Reply:
Yes, it will ignore the abnormal aggregated APIService because we cannot fetch the resources of the abnormal aggregated api

Reply:
When will the addition of the aggregate APIService Group to our discoveryCache be triggered.

Reply:
When the status of the `Available` condition changes to True, it usually means that the service can be accessed normally and will be added to the discoveryCache